onvif/src/discovery/mod.rs

```diff
@@ -150,22 +151,37 @@ async fn get_responding_addr<F, Fut>(
     mut check_addr: F,
 ) -> Option<Device>
 where
-    F: FnMut(Url) -> Fut,
+    F: FnMut(Url) -> Fut + Copy,
     Fut: Future<Output = bool>,
 {
-    for probe_match in &envelope.body.probe_matches.probe_match {
-        for url in probe_match.x_addrs() {
-            if check_addr(url.clone()).await {
-                debug!("Responding addr: {:?}", url);
-                return Some(Device {
-                    url,
-                    name: probe_match.name(),
-                });
-            }
-        }
-    }
-
-    None
+    envelope
+        .body
+        .probe_matches
+        .probe_match
+        .iter()
+        .map(|probe_match| async move {
+            probe_match
+                .x_addrs()
+                .iter()
+                .map(|url| async move {
+                    if check_addr(url.clone()).await {
+                        debug!("Responding addr: {:?}", url);
+                        Some(Device {
+                            name: probe_match.name(),
+                            url: url.clone(),
+                        })
+                    } else {
+                        None
+                    }
+                })
+                .collect::<FuturesUnordered<_>>()
+                .next()
+                .await
+        })
```

Comment:
Could we do sth. like this instead?

```rust
// use std::iter;
.flat_map(|probe_match| {
    iter::zip(probe_match.x_addrs(), iter::repeat(probe_matches.name())
})
.map(|(url, name)| { /* if check_addr(... */ })
```

Reply:
Oh, also that `if` / `else` could be shortened using [bool::then](https://doc.rust-lang.org/std/primitive.bool.html#method.then) instead.

Reply:
Sorry, had to rewrite that. I just realized that we cannot first `flat_map` everything because we have to return exactly one responding address per camera. When we `flat_map`, we can get multiple `responding address` + `name` pairs per camera.

`bool::then` is nice, I always forget that it is stable already.

Reply:
`zip` gives me an error:
```
error[E0658]: use of unstable library feature 'iter_zip'
   --> onvif/src/discovery/mod.rs:169:13
    |
169 |             iter::zip(probe_match.x_addrs(), iter::repeat(probe_match.name()))
    |             ^^^^^^^^^
    |
    = note: see issue #83574 <https://github.com/rust-lang/rust/issues/83574> for more information
    = help: add `#![feature(iter_zip)]` to the crate attributes to enable
```

Can use `itertools` though.

Reply:
You can just do `.x_addrs().into_iter().zip(iter::repeat(/* ... */` instead. Anyways, I only now see that there's `.next()` calls and am a bit confused. Maybe the next revision of this will be clearer to me 😅

Reply:
The thing about `next` is that we want to get only result from the first future executed to completion discarding the others. `FuturesUnordered` allows to execute futures concurrently without order.

Reply:
> Sorry, had to rewrite that. I just realized that we cannot first `flat_map` everything because we have to return exactly one responding address per camera. When we `flat_map`, we can get multiple `responding address` + `name` pairs per camera.
> 
> `bool::then` is nice, I always forget that it is stable already.

Ah, wait, this is wrong :facepalm: A bit lost in these probe matches. `flat_map` should be fine as this function processes answer from exactly one camera, so the final `next()` is enough.

Reply:
Alright, now it should be fine. Please have a look.